server/data_provider.rb

```diff
@@ -224,42 +177,142 @@ def dataProviderToDoMnemeLMS(uniqname)
+    #!config_hash[:data_provider_file_directory].nil? ? configureFileProvider(config_hash) : configureEsbProvider(config_hash)
+  end
 
-  def dataProviderTerms(uniqname)
+  ### Hide any decisions about which ctools connection method to use.
+  def configureCToolsProvider(config_hash)
 
-    logger.debug "#{self.class.to_s}:#{__method__}: #{__LINE__}: DataProviderTerms uniqname: #{uniqname}"
+    logger.error "@@@@@@@@@@@@@ must specify one CTools data provider" unless verifyExactlyOneProvider(config_hash,[:data_provider_file_directory,:ctools_http_application_name])
+    configureCToolsFileProvider(config_hash) unless (config_hash[:data_provider_file_directory].nil?)
+    configureCToolsHTTPProvider(config_hash) unless (config_hash[:ctools_http_application_name].nil?)
 
-    dataProviderInit
+  end
 
-    terms = @fileTerms.(uniqname) if @useFileProvider
-    terms = @esbTerms.(uniqname) if @useEsbProvider
+  ## These methods delegate configuration to the provider.
+  def configureCToolsFileProvider(config_hash)
+    logger.debug "#{self.class.to_s}:#{__method__}: #{__LINE__}: call configure"
+    # Assign or merge hash?
+    @ctoolsHash = initConfigureCToolsFileProvider(config_hash)
+  end
 
-    logIfUnavailable(terms, "terms: user: #{uniqname}")
+  ## These methods delegate configuration to the provider.
+  def configureCanvasFileProvider(config_hash)
+    logger.debug "#{self.class.to_s}:#{__method__}: #{__LINE__}: call configure"
+    # Assign or merge hash?
+    @canvasHash = initConfigureCanvasFileProvider(config_hash)
+  end
 
-    terms
+  ### Hide any decisions about which canvas provider to use.
+  def configureCanvasProvider(config_hash)
+    logger.error "@@@@@@@@@@@@@ must specify one canvas data provider" unless verifyExactlyOneProvider(config_hash,[:data_provider_file_directory,:canvas_esb_application_name,:canvas_http_application_name])
+    configureCanvasFileProvider(config_hash) unless (config_hash[:data_provider_file_directory].nil?)
+    configureCanvasESBProvider(config_hash) unless (config_hash[:canvas_esb_application_name].nil?)
+    configureCanvasHTTPProvider(config_hash) unless (config_hash[:canvas_http_application_name].nil?)
   end
 
-  def dataProviderCourse(uniqname, termid)
+  ## Configuration binds implementation specific information to hide details and allow interchangable calling of
+  ## any implementation regardless of configuration details.
 
-    dataProviderInit
+  ## TODO: should apply to all data sources including ToDo
+  def configureFileProvider(config_hash)
+    dpf_dir = config_hash[:data_provider_file_directory]
+    logger.debug "#{self.class.to_s}:#{__method__}:#{__LINE__}: configure provider file: directory: [#{dpf_dir}]"
 
-    courses = @fileCourses.(uniqname, termid) if @useFileProvider
-    courses = @esbCourses.(uniqname, termid) if @useEsbProvider
+    @useFileProvider = true
+    @fileTerms = Proc.new { |uniqname| dataProviderFileTerms("#{dpf_dir}/terms", uniqname) }
```

Comment:
Don't understand why the @fileterms and @fileCourses are dependent on uniqname as shown in the param list.